File: src/de/espend/idea/php/annotation/reference/AnnotationPropertyValueReferenceContributor.java
package de.espend.idea.php.annotation.reference;

import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.*;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.ProcessingContext;
import com.jetbrains.php.lang.documentation.phpdoc.lexer.PhpDocTokenTypes;
import com.jetbrains.php.lang.documentation.phpdoc.psi.tags.PhpDocTag;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import de.espend.idea.php.annotation.AnnotationPropertyParameter;
import de.espend.idea.php.annotation.PhpAnnotationReferencesProvider;
import de.espend.idea.php.annotation.pattern.AnnotationPattern;
import de.espend.idea.php.annotation.reference.parameter.ReferencesByElementParameter;
import de.espend.idea.php.annotation.util.AnnotationUtil;
import de.espend.idea.php.annotation.util.PhpElementsUtil;
import de.espend.idea.php.annotation.util.PluginUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;

public class AnnotationPropertyValueReferenceContributor extends PsiReferenceContributor {

    @Override
    public void registerReferenceProviders(PsiReferenceRegistrar psiReferenceRegistrar) {
        psiReferenceRegistrar.registerReferenceProvider(AnnotationPattern.getDefaultPropertyValueString(), new PropertyValueDefaultReferences());
        psiReferenceRegistrar.registerReferenceProvider(AnnotationPattern.getPropertyValueString(), new PropertyValueReferences());
    }

    private class PropertyValueDefaultReferences extends PsiReferenceProvider {

        @NotNull
        @Override
        public PsiReference[] getReferencesByElement(@NotNull PsiElement psiElement, @NotNull ProcessingContext processingContext) {

            if(!PluginUtil.isEnabled(psiElement)) {
                return new PsiReference[0];
            }

            PhpClass phpClass = getValidAnnotationClass(psiElement);
            if(phpClass == null) {
                return new PsiReference[0];
            }

            AnnotationPropertyParameter annotationPropertyParameter = new AnnotationPropertyParameter(psiElement, phpClass, AnnotationPropertyParameter.Type.DEFAULT);
            return addPsiReferences(psiElement, processingContext, annotationPropertyParameter);

        }
    }

    private class PropertyValueReferences extends PsiReferenceProvider {

        @NotNull
        @Override
        public PsiReference[] getReferencesByElement(@NotNull PsiElement psiElement, @NotNull ProcessingContext processingContext) {

            if(!PluginUtil.isEnabled(psiElement)) {
                return new PsiReference[0];
            }

            PhpClass phpClass = getValidAnnotationClass(psiElement);
            if(phpClass == null) {
                return new PsiReference[0];
            }

            PsiElement propertyName = PhpElementsUtil.getPrevSiblingOfPatternMatch(psiElement, PlatformPatterns.psiElement(PhpDocTokenTypes.DOC_IDENTIFIER));
            if(propertyName == null) {
                return new PsiReference[0];
            }

            AnnotationPropertyParameter annotationPropertyParameter = new AnnotationPropertyParameter(psiElement, phpClass, propertyName.getText(), AnnotationPropertyParameter.Type.STRING);
            return addPsiReferences(psiElement, processingContext, annotationPropertyParameter);

        }
    }

    @Nullable
    private PhpClass getValidAnnotationClass(PsiElement psiElement) {
        PhpDocTag phpDocTag = PsiTreeUtil.getParentOfType(psiElement, PhpDocTag.class);
        if(phpDocTag == null) {
            return null;
        }

        return AnnotationUtil.getAnnotationReference(phpDocTag);
    }

    private PsiReference[] addPsiReferences(PsiElement psiElement, ProcessingContext processingContext, AnnotationPropertyParameter annotationPropertyParameter) {
        ArrayList<PsiReference> psiReferences = new ArrayList<PsiReference>();

        ReferencesByElementParameter referencesByElementParameter = new ReferencesByElementParameter(psiElement, processingContext);

        for(PhpAnnotationReferencesProvider phpAnnotationExtension : AnnotationUtil.EXTENSION_POINT_REFERENCES.getExtensions()) {
            PsiReference[] references = phpAnnotationExtension.getPropertyReferences(annotationPropertyParameter, referencesByElementParameter);
            if(references != null && references.length > 0) {
                psiReferences.addAll(Arrays.asList(references));
            }
        }

        return psiReferences.toArray(new PsiReference[psiReferences.size()]);
    }

}

File: src/main/kotlin/de/espend/idea/php/annotation/reference/AnnotationPropertyValueReferenceContributor.kt
package de.espend.idea.php.annotation.reference

import com.intellij.patterns.ElementPattern
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReference
import com.intellij.psi.PsiReferenceContributor
import com.intellij.psi.PsiReferenceProvider
import com.intellij.psi.PsiReferenceRegistrar
import com.intellij.psi.PsiWhiteSpace
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.ProcessingContext
import com.jetbrains.php.lang.documentation.phpdoc.lexer.PhpDocTokenTypes
import com.jetbrains.php.lang.documentation.phpdoc.psi.tags.PhpDocTag
import com.jetbrains.php.lang.lexer.PhpTokenTypes
import com.jetbrains.php.lang.psi.PhpPsiUtil
import com.jetbrains.php.lang.psi.elements.PhpAttribute
import com.jetbrains.php.lang.psi.elements.PhpClass
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import de.espend.idea.php.annotation.extension.parameter.AnnotationPropertyParameter
import de.espend.idea.php.annotation.extension.parameter.PhpAnnotationReferenceProviderParameter
import de.espend.idea.php.annotation.pattern.AnnotationPattern
import de.espend.idea.php.annotation.util.AnnotationUtil
import de.espend.idea.php.annotation.util.PhpElementsUtil
import org.apache.commons.lang3.StringUtils

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
class AnnotationPropertyValueReferenceContributor : PsiReferenceContributor() {
    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {
        registrar.registerReferenceProvider(
            AnnotationPattern.getDefaultPropertyValueString(),
            PropertyValueDefaultReferences(),
        )
        registrar.registerReferenceProvider(
            AnnotationPattern.getPropertyValueString(),
            PropertyValueReferences(),
        )

        // #[Route('/path', name: '<caret>')]
        registrar.registerReferenceProvider(
            AnnotationPattern.getAttributesValueReferencesPattern(),
            AttributeValueReferences(),
        )

        // #[Route('<caret>')]
        registrar.registerReferenceProvider(
            AnnotationPattern.getAttributesDefaultPattern(),
            AttributeDefaultReferences(),
        )
    }

    /**
     * '@Template("foo.twig.html")'
     * '@Service("foo")'
     */
    private inner class PropertyValueDefaultReferences : PsiReferenceProvider() {
        override fun getReferencesByElement(
            element: PsiElement,
            context: ProcessingContext,
        ): Array<PsiReference> {
            val phpClass = getValidAnnotationClass(element) ?: return emptyArray()
            val property = AnnotationPropertyParameter(element, phpClass, AnnotationPropertyParameter.Type.DEFAULT)
            return addPsiReferences(element, context, property)
        }
    }

    /**
     * '@Template(name="foo.twig.html")'
     */
    private inner class PropertyValueReferences : PsiReferenceProvider() {
        override fun getReferencesByElement(
            element: PsiElement,
            context: ProcessingContext,
        ): Array<PsiReference> {
            val phpClass = getValidAnnotationClass(element) ?: return emptyArray()
            val propertyName = PhpElementsUtil.getPrevSiblingOfPatternMatch(element, DOC_IDENTIFIER_PATTERN)
                ?: return emptyArray()
            val property = AnnotationPropertyParameter(
                element,
                phpClass,
                propertyName.text,
                AnnotationPropertyParameter.Type.PROPERTY_VALUE,
            )
            return addPsiReferences(element, context, property)
        }
    }

    /**
     * `#[Route('/path', name: '<caret>')]`
     */
    private inner class AttributeValueReferences : PsiReferenceProvider() {
        override fun getReferencesByElement(
            element: PsiElement,
            context: ProcessingContext,
        ): Array<PsiReference> {
            val phpClass = getAttributeClass(element) ?: return emptyArray()
            val attributeNamePsi = PhpPsiUtil.getPrevSibling(
                element,
                { sibling ->
                    sibling is PsiWhiteSpace || sibling.node.elementType === PhpTokenTypes.opCOLON
                },
            )
            if (attributeNamePsi == null || attributeNamePsi.node.elementType !== PhpTokenTypes.IDENTIFIER) {
                return emptyArray()
            }

            val attributeName = attributeNamePsi.text
            if (StringUtils.isBlank(attributeName)) {
                return emptyArray()
            }

            val property = AnnotationPropertyParameter(
                element,
                phpClass,
                attributeName,
                AnnotationPropertyParameter.Type.PROPERTY_VALUE,
            )
            return addPsiReferences(element, context, property)
        }
    }

    /**
     * `#[Route('<caret>')]`
     */
    private inner class AttributeDefaultReferences : PsiReferenceProvider() {
        override fun getReferencesByElement(
            element: PsiElement,
            context: ProcessingContext,
        ): Array<PsiReference> {
            val phpClass = getAttributeClass(element) ?: return emptyArray()
            val property = AnnotationPropertyParameter(element, phpClass, AnnotationPropertyParameter.Type.DEFAULT)
            return addPsiReferences(element, context, property)
        }
    }

    private fun getValidAnnotationClass(element: PsiElement): PhpClass? {
        val phpDocTag = PsiTreeUtil.getParentOfType(element, PhpDocTag::class.java) ?: return null
        return AnnotationUtil.getAnnotationReference(phpDocTag)
    }

    private fun addPsiReferences(
        element: PsiElement,
        context: ProcessingContext,
        property: AnnotationPropertyParameter,
    ): Array<PsiReference> {
        val references = ArrayList<PsiReference>()
        val parameter = PhpAnnotationReferenceProviderParameter(element, context)

        for (extension in AnnotationUtil.EXTENSION_POINT_REFERENCES.extensions) {
            val extensionReferences = extension.getPropertyReferences(property, parameter)
            if (!extensionReferences.isNullOrEmpty()) {
                references.addAll(extensionReferences.filterNotNull())
            }
        }

        return references.toTypedArray()
    }

}

private val DOC_IDENTIFIER_PATTERN: ElementPattern<PsiElement> =
    PlatformPatterns.psiElement(PhpDocTokenTypes.DOC_IDENTIFIER)

private fun getAttributeClass(element: PsiElement): PhpClass? {
    if (element !is StringLiteralExpression) {
        return null
    }

    val phpAttribute = PsiTreeUtil.getParentOfType(element, PhpAttribute::class.java) ?: return null
    val fqn = phpAttribute.fqn ?: return null
    return PhpElementsUtil.getClassInterface(element.project, fqn)
}
